syntax = "proto3";

package ai.stigmer.commons.apiresource;

import "ai/stigmer/commons/apiresource/apiresourcekind/api_resource_kind.proto";
import "ai/stigmer/commons/apiresource/enum.proto";
import "ai/stigmer/commons/rpc/pagination.proto";
import "buf/validate/validate.proto";

// Wrapper for api-resource-id
message ApiResourceId {
  string value = 1 [(buf.validate.field).required = true];
}

// Input for rpc to delete or preview delete for any api resource
message ApiResourceDeleteInput {
  // ID of the resource
  string resource_id = 1 [(buf.validate.field).required = true];
  // A descriptive message explaining the reason for deleting the resource. This is used for audit purposes.
  // For example, "Deleting as it is no longer needed".
  string version_message = 2;
  // Flag for api-resources that require force deletion override from the clients
  bool force = 3;
}

// Input for requests that need api-resource org and slug
message ApiResourceByOrgBySlugRequest {
  // ID of the organization
  string org = 1 [(buf.validate.field).required = true];
  // Slug of the resource
  string slug = 2 [(buf.validate.field).required = true];
}

// Input for find (list) operations with pagination
message FindApiResourcesRequest {
  // Organization ID to filter by
  string org = 1 [(buf.validate.field).required = true];
  // API resource kind (optional filter)
  string kind = 3;
  // Pagination info
  ai.stigmer.commons.rpc.PageInfo page = 4;
  // Page number (1-indexed) - deprecated, use page.num
  int32 page_number = 5;
  // Number of items per page (default: 20, max: 100) - deprecated, use page.size
  int32 page_size = 6;
}

// Generic reference to any API resource by scope, org, and name.
// Used across resources to reference other resources (e.g., Environment, Agent, etc.)
message ApiResourceReference {
  // Owner scope of the referenced resource
  ApiResourceOwnerScope scope = 1 [(buf.validate.field).enum.defined_only = true];

  // Organization ID (required if scope = organization)
  string org = 2;

  ai.stigmer.commons.apiresource.apiresourcekind.ApiResourceKind kind = 3;

  // Resource slug (user-friendly name, not ID)
  string slug = 4;

  // Version of the resource (optional, only applicable to versioned resources like Skills).
  // Supports three formats:
  // 1. Empty/unset → Resolves to "latest" (most recent version)
  // 2. Tag name → Resolves to version with this tag (e.g., "stable", "v1.0")
  // 3. Exact hash → Immutable reference to specific version (e.g., "abc123...")
  //
  // Default behavior: Empty means "latest" (current version).
  // This field is ignored for non-versioned resources.
  //
  // Examples:
  // - version: ""           → Use latest version
  // - version: "latest"     → Use latest version (explicit)
  // - version: "stable"     → Use version tagged as "stable"
  // - version: "v1.0"       → Use version tagged as "v1.0"
  // - version: "abc123..."  → Use exact version with this hash (immutable)
  string version = 5 [(buf.validate.field).string.pattern = "^$|^latest$|^[a-zA-Z0-9._-]+$|^[a-f0-9]{64}$"];
}
